gobblin-service/src/main/java/org/apache/gobblin/service/modules/flowgraph/Dag.java

```diff
@@ -0,0 +1,148 @@
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.gobblin.service.modules.flowgraph;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+import org.apache.gobblin.annotation.Alpha;
+
+import com.google.common.collect.Lists;
+
+import lombok.Getter;
+
+
+/**
+ * An implementation of Dag. Assumes that nodes have unique values. Nodes with duplicate values will produce
+ * unpredictable behavior.
+ */
+@Alpha
+@Getter
+public class Dag<T> {
+  private List<DagNode<T>> startNodes;
+  private List<DagNode<T>> endNodes;
+  // Map to maintain parent to children mapping.
+  private Map<DagNode, List<DagNode<T>>> parentChildMap;
+  private List<DagNode<T>> nodes;
+
+  public Dag(List<DagNode<T>> dagNodes) {
+    this.nodes = dagNodes;
+    //Build dag
+    this.build();
+  }
+
+  /**
+   * Constructs the dag from the Node list.
+   */
+  private void build() {
+    this.startNodes = new ArrayList<>();
+    this.endNodes = new ArrayList<>();
+    this.parentChildMap = new HashMap<>();
+    for (DagNode node : this.nodes) {
+      //If a Node has no parent Node, add it to the list of start Nodes
+      if (node.getParentNodes() == null) {
+        this.startNodes.add(node);
+      } else {
+        List<DagNode> parentNodeList = node.getParentNodes();
+        for (DagNode parentNode : parentNodeList) {
+          if (parentChildMap.containsKey(parentNode)) {
+            parentChildMap.get(parentNode).add(node);
+          } else {
+            parentChildMap.put(parentNode, Lists.newArrayList(node));
+          }
+        }
+      }
+    }
+    //Iterate over all the Nodes and add a Node to the list of endNodes if it is not present in the parentChildMap
+    for (DagNode node : this.nodes) {
+      if (!parentChildMap.containsKey(node)) {
+        this.endNodes.add(node);
+      }
+    }
+  }
+
+  public List<DagNode<T>> getChildren(DagNode node) {
+    return parentChildMap.getOrDefault(node, null);
+  }
+
+  public List<DagNode<T>> getParents(DagNode node) {
+    return node.parentNodes;
+  }
+
+  /**
+   * Concatenate two dags together. Join the "other" dag to "this" dag and return "this" dag.
+   * @return the concatenated dag
+   */
+  public Dag<T> concatenate(Dag<T> other) throws IOException {
+    if (other == null) {
+      return this;
+    }
+    for (DagNode node : this.endNodes) {
```

Comment:
why not a similar logic as line 64? why proactively put all endNodes as parent?

Reply:
The Dag concatenation is ensuring that all the jobs of the previous edge (which may have multiple end nodes) are completed before starting any job of this edge. The logic is adding the endNodes of previous edge as dependencies of the startNodes of the next edge.